core/domain/story_domain_test.py

```diff
@@ -1114,79 +1118,438 @@ def test_get_all_linked_exp_ids(self) -> None:
         ]
+
+        self.story.update_node_exploration_id('node_1', 'exp_1')
+
         self.assertEqual(
             self.story.story_contents.get_all_linked_exp_ids(),
             ['exp_1', 'exp_2'])
+
+    def test_update_node_exploration_id_raises_error_when_introducing_duplicate_exploration_id( # pylint: disable=line-too-long
+        self
+    ) -> None:
+        self.story.story_contents.next_node_id = 'node_3'
+        node_1: story_domain.StoryNodeDict = {
+            'id': 'node_1',
+            'thumbnail_filename': 'image.svg',
+            'thumbnail_bg_color': constants.ALLOWED_THUMBNAIL_BG_COLORS[
+                'chapter'][0],
+            'thumbnail_size_in_bytes': 21131,
+            'title': 'Title 1',
+            'description': 'Description 1',
+            'destination_node_ids': ['node_3'],
+            'acquired_skill_ids': [],
+            'prerequisite_skill_ids': [],
+            'outline': '',
+            'outline_is_finalized': False,
+            'exploration_id': 'exp_1',
+            'status': 'Published',
+            'planned_publication_date_msecs': 100,
+            'last_modified_msecs': 100,
+            'first_publication_date_msecs': 100,
+            'unpublishing_reason': None
+        }
+        node_2: story_domain.StoryNodeDict = {
+            'id': 'node_2',
+            'thumbnail_filename': 'image.svg',
+            'thumbnail_bg_color': constants.ALLOWED_THUMBNAIL_BG_COLORS[
+                'chapter'][0],
+            'thumbnail_size_in_bytes': 21131,
+            'title': 'Title 2',
+            'description': 'Description 2',
+            'destination_node_ids': ['node_1'],
+            'acquired_skill_ids': [],
+            'prerequisite_skill_ids': [],
+            'outline': '',
+            'outline_is_finalized': False,
+            'exploration_id': 'exp_2',
+            'status': 'Published',
+            'planned_publication_date_msecs': 100,
+            'last_modified_msecs': 100,
+            'first_publication_date_msecs': 100,
+            'unpublishing_reason': None
+        }
+        self.story.story_contents.initial_node_id = 'node_2'
         self.story.story_contents.nodes = [
             story_domain.StoryNode.from_dict(node_1),
-            story_domain.StoryNode.from_dict(node_2),
-            story_domain.StoryNode.from_dict(node_4),
+            story_domain.StoryNode.from_dict(node_2)
         ]
-        self.assertEqual(
-            self.story.story_contents.get_all_linked_exp_ids(),
-            ['exp_1', 'exp_2'])
-        self.story.story_contents.nodes.append(
-            story_domain.StoryNode.from_dict(node_3))
-        self.assertEqual(
-            self.story.story_contents.get_all_linked_exp_ids(),
-            ['exp_1', 'exp_2', 'exp_3'])
+
         with self.assertRaisesRegex(
-            ValueError, 'A node with exploration id exp_3 already exists.'
+             ValueError, 'A node with exploration id exp_2 already exists.'
```

Comment:
Drop the single added space before ValueError

Reply:
Done.